<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="enONE">
<title>enONE_tutorial • enONE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="enONE_tutorial">
<meta property="og:description" content="enONE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">enONE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-vignettes">Vignettes</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-vignettes">
    <a class="dropdown-item" href="../articles/enONE_tutorial.html">enONE Tutorial</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../../">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/thereallda/enONE">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>enONE_tutorial</h1>
            
      
      
      <div class="d-none name"><code>enONE_tutorial.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The hub metabolite, nicotinamide adenine dinucleotide (NAD), can be
used as an initiating nucleotide in RNA transcription to result in
NAD-capped RNAs (NAD-RNAs). Epitranscriptome-wide profiling of NAD-RNAs
involves chemo-enzymatic labeling and affinity-based enrichment; yet
currently available computational analysis cannot adequately remove
variations inherently linked with capture procedures. Here, we propose a
spike-in-based normalization and data-driven evaluation framework,
<code>enONE</code>, for the omic-level analysis of NAD-capped RNAs.</p>
<p><code>enONE</code> package is implemented in R and publicly available
at <a href="https://github.com/thereallda/enONE" class="external-link uri">https://github.com/thereallda/enONE</a>.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>For this tutorial, we will demonstrate the <code>enONE</code>
workflow by using a NAD-RNA-seq data from human peripheral blood
mononuclear cells (PBMCs).</p>
<p>Notably, we included three types of spike-in RNAs:</p>
<ol style="list-style-type: decimal">
<li>Total RNAs from <em>Drosophila melanogaster</em>, an invertebrate
model organism with well-annotated genome sequence, for estimating the
unwanted variation;</li>
<li>Synthetic RNAs, consisting of 5% NAD- relative to m7G-capped forms,
were used to determine the capture sensitivity;</li>
<li>Synthetic RNAs, with 100% m7G-capped forms, were used to determine
the capture specificity (Figure1 below).</li>
</ol>
<p><img src="images/Figure1.png"></p>
<p><em>Figure 1: Schematic workflow for total RNAs from PBMCs and three
sets of spike-ins.</em></p>
<p>We start by reading in the data, including count matrix and metadata.
It is recommended to provide the assay group of samples with the
<code>assay</code> column in metadata.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thereallda.github.io/enONE/">enONE</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read in metadata and counts matrix</span></span>
<span><span class="va">counts_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"data/Counts.csv"</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"data/metadata.csv"</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span></span>
<span><span class="co">#&gt;   id sample.id condition batch assay</span></span>
<span><span class="co">#&gt; 1 D1        Y2   Y.Input     D Input</span></span>
<span><span class="co">#&gt; 2 D2       Y10   Y.Input     D Input</span></span>
<span><span class="co">#&gt; 3 D3       Y14   Y.Input     D Input</span></span>
<span><span class="co">#&gt; 4 D4       Y16   Y.Input     D Input</span></span>
<span><span class="co">#&gt; 5 D5       M13   M.Input     D Input</span></span>
<span><span class="co">#&gt; 6 D6       M14   M.Input     D Input</span></span></code></pre></div>
<p>Next, we use the count matrix to create a <code>Enone</code>
object.</p>
<p>The <code>Enone</code> object serves as a container that contains
both data (raw and normalized counts data) and analysis (e.g.,
normalization performance evaluation score and enrichment results) for a
NAD-RNA-seq dataset.</p>
<p>When constructing <code>Enone</code> object, we can provide following
arguments:</p>
<ul>
<li><p>Prefix of spike-in genes
(<code>spike.in.prefix = "^FB"</code>);</p></li>
<li><p>The id of synthetic spike-in
(<code>synthetic.id = c("Syn1", "Syn2")</code>), same with row names in
<code>counts_mat</code>;</p></li>
<li><p>The id of input (<code>input.id = "Input"</code>) and enrichment
(<code>enrich.id = "Enrich"</code>), same as the assay column.</p></li>
</ul>
<blockquote>
<p>Here, “Syn1” represent the spike-in with 5% NAD-RNA and “Syn2” is the
spike-in with 100% m7G-RNA.</p>
<p><strong>Notably, <code>batch.group</code> , <code>synthetic.id</code>
are optional.</strong></p>
</blockquote>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prefix of Drosophila spike-in genes</span></span>
<span><span class="va">spikeInPrefix</span> <span class="op">&lt;-</span> <span class="st">"^FB"</span></span>
<span></span>
<span><span class="co"># create Enone</span></span>
<span><span class="va">Enone</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createEnone.html">createEnone</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">counts_mat</span>,</span>
<span>                     bio.group <span class="op">=</span> <span class="va">meta</span><span class="op">$</span><span class="va">condition</span>,</span>
<span>                     enrich.group <span class="op">=</span> <span class="va">meta</span><span class="op">$</span><span class="va">assay</span>,</span>
<span>                     batch.group <span class="op">=</span> <span class="va">meta</span><span class="op">$</span><span class="va">batch</span>,</span>
<span>                     spike.in.prefix <span class="op">=</span> <span class="va">spikeInPrefix</span>,</span>
<span>                     synthetic.id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Syn1"</span>, <span class="st">"Syn2"</span><span class="op">)</span>, </span>
<span>                     input.id <span class="op">=</span> <span class="st">"Input"</span>,</span>
<span>                     enrich.id <span class="op">=</span> <span class="st">"Enrich"</span></span>
<span>                     <span class="op">)</span></span>
<span><span class="va">Enone</span></span>
<span><span class="co">#&gt; class: Enone </span></span>
<span><span class="co">#&gt; dim: 76290 62 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): ''</span></span>
<span><span class="co">#&gt; rownames(76290): ENSG00000223972 ENSG00000227232 ... Syn1 Syn2</span></span>
<span><span class="co">#&gt; rowData names(3): GeneID SpikeIn Synthetic</span></span>
<span><span class="co">#&gt; colnames(62): D1 D2 ... F22 F24</span></span>
<span><span class="co">#&gt; colData names(5): id condition enrich replicate batch</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="standard-workflow">Standard workflow<a class="anchor" aria-label="anchor" href="#standard-workflow"></a>
</h2>
<p>The <code>enONE</code> workflow consists of four steps:</p>
<ol style="list-style-type: decimal">
<li>Quality control;</li>
<li>Gene set selection;</li>
<li>Normalization procedures;</li>
<li>Normalization performance assessment.</li>
</ol>
<div class="section level3">
<h3 id="quality-control">Quality Control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h3>
<p><code>enONE</code> initiate with quality control step to filter
lowly-expressed genes. This step is performed with
<code>FilterLowExprGene</code> by keeping genes with at least
<code>min.count</code> in at least n samples. n is determined by the
smallest group sample size specifying in <code>group</code> .</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Enone</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterLowExprGene.html">FilterLowExprGene</a></span><span class="op">(</span><span class="va">Enone</span>, group <span class="op">=</span> <span class="va">Enone</span><span class="op">$</span><span class="va">condition</span>, min.count <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">Enone</span></span>
<span><span class="co">#&gt; class: Enone </span></span>
<span><span class="co">#&gt; dim: 27661 62 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): ''</span></span>
<span><span class="co">#&gt; rownames(27661): ENSG00000227232 ENSG00000268903 ... Syn1 Syn2</span></span>
<span><span class="co">#&gt; rowData names(3): GeneID SpikeIn Synthetic</span></span>
<span><span class="co">#&gt; colnames(62): D1 D2 ... F22 F24</span></span>
<span><span class="co">#&gt; colData names(5): id condition enrich replicate batch</span></span></code></pre></div>
<p>Additionally, outliers can be assessed and further removed (if
<code>return=TRUE</code>) by <code>OutlierTest</code>. Since no sample
is flagged as outlier, we will use all samples in following
analysis.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## ronser"s test for outlier assessment</span></span>
<span><span class="fu"><a href="../reference/OutlierTest.html">OutlierTest</a></span><span class="op">(</span><span class="va">Enone</span>, return<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rosner's outlier test</span></span>
<span><span class="co">#&gt;   i        Mean.i     SD.i    Value Obs.Num    R.i+1 lambda.i+1 Outlier</span></span>
<span><span class="co">#&gt; 1 0  1.158348e-16 51.77231 80.37523      15 1.552475   3.212165   FALSE</span></span>
<span><span class="co">#&gt; 2 1 -1.317627e+00 51.14304 78.20437      19 1.554894   3.205977   FALSE</span></span>
<span><span class="co">#&gt; 3 2 -2.642993e+00 50.50717 70.38990      41 1.445991   3.199662   FALSE</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="running-enone">Running enONE<a class="anchor" aria-label="anchor" href="#running-enone"></a>
</h3>
<p>The main <code>enONE</code> function can be used to perform gene
selection, normalization, and evaluation. This function return selected
gene set in <code>rowData</code>, normalized count matrix in
<code>counts</code> slot (when <code>return.norm=TRUE</code>),
normalization factors (<code>enone_factor</code> slot), evaluation
metrics and scores in <code>enone_metrics</code> and
<code>enone_scores</code> slots, respectively. Detailed of these steps
are described bellow.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Enone</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enONE.html">enONE</a></span><span class="op">(</span><span class="va">Enone</span>, </span>
<span>               scaling.method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TC"</span>, <span class="st">"UQ"</span>, <span class="st">"TMM"</span>, <span class="st">"DESeq"</span>, <span class="st">"PossionSeq"</span><span class="op">)</span>,</span>
<span>               ruv.norm <span class="op">=</span> <span class="cn">TRUE</span>, ruv.k <span class="op">=</span> <span class="fl">3</span>,</span>
<span>               eval.pam.k <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">6</span>, eval.pc.n <span class="op">=</span> <span class="fl">3</span>, </span>
<span>               return.norm <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>               <span class="op">)</span></span>
<span><span class="co">#&gt; Gene set selection for normalization and assessment...</span></span>
<span><span class="co">#&gt; - The number of negative control genes for normalization: 1000 </span></span>
<span><span class="co">#&gt; - Estimate dispersion &amp; Fit GLM... </span></span>
<span><span class="co">#&gt; - Testing differential genes... </span></span>
<span><span class="co">#&gt; - The number of positive evaluation genes: 500 </span></span>
<span><span class="co">#&gt; - Estimate dispersion &amp; Fit GLM... </span></span>
<span><span class="co">#&gt; - Testing differential genes... </span></span>
<span><span class="co">#&gt; - The number of negative evaluation genes: 500 </span></span>
<span><span class="co">#&gt; - Estimate dispersion &amp; Fit GLM... </span></span>
<span><span class="co">#&gt; - Testing differential genes... </span></span>
<span><span class="co">#&gt; Apply normalization...</span></span>
<span><span class="co">#&gt; - Scaling... </span></span>
<span><span class="co">#&gt; - Regression-based normalization... </span></span>
<span><span class="co">#&gt; Perform assessment...</span></span></code></pre></div>
<div class="section level4">
<h4 id="gene-set-selection">Gene set Selection<a class="anchor" aria-label="anchor" href="#gene-set-selection"></a>
</h4>
<p>For gene set selection, enONE defined three sets of control genes,
including:</p>
<ul>
<li><p>Negative Control (<code>NegControl</code>): By default,
<code>enONE</code> define the 1,000 least significantly enriched genes
in <em>Drosophila</em> spike-ins (or other RNA spike-in from exogenous
organism), ranked by FDR, as the negative controls for adjustment of the
unwanted variations.</p></li>
<li><p>Negative Evaluation (<code>NegEvaluation</code>): By default,
<code>enONE</code> define the 500 least significantly varied genes in
samples of interest, ranked by FDR, as negative evaluation genes for
evaluation of the unwanted variations.</p></li>
<li><p>Positive Evaluation (<code>PosEvaluation</code>): By default,
<code>enONE</code> define the 500 most significantly enriched genes in
samples of interest, ranked by FDR, as positive evaluation genes for
evaluation of the wanted variations.</p></li>
</ul>
<p>Gene set selection can either be automatically defined in
<code>enONE</code> function with <code>auto=TRUE</code> parameter
(default), or be provided in
<code>neg.control, pos.eval, neg.eval</code> parameters,
respectively.</p>
<p>Selected gene sets can be assessed by <code>getGeneSet</code> with
the name of gene set (i.e., <code>"NegControl"</code>,
<code>"NegEvaluation"</code>, <code>"PosEvaluation"</code>).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getGeneSet.html">getGeneSet</a></span><span class="op">(</span><span class="va">Enone</span>, name <span class="op">=</span> <span class="st">"NegControl"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "FBgn0031247" "FBgn0031255" "FBgn0004583" "FBgn0031324" "FBgn0026397"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h4>
<p><code>enONE</code> implements global scaling and regression-based
methods for the generation of normalization procedures.</p>
<p>For the global scaling normalization procedures, five different
scaling procedures are implemented, including</p>
<ol style="list-style-type: decimal">
<li><p>Total Count (TC);</p></li>
<li><p>Upper-Quartile (UQ);</p></li>
<li><p>Trimmed Mean of M Values (TMM);</p></li>
<li><p>DESeq;</p></li>
<li>
<p>PossionSeq.</p>
<p>By default, enONE uses all scaling methods, while user can perform
with selected scaling procedures in <code>scaling.method</code>
parameter.</p>
</li>
</ol>
<p>For the regression-based procedures, <code>enONE</code> use three
variants of RUV to estimate the factors of unwanted variation,
including</p>
<ol style="list-style-type: decimal">
<li><p>RUVg;</p></li>
<li><p>RUVs;</p></li>
<li>
<p>RUVse.</p>
<p>For instance, you can perform RUV with selected the first two
unwanted factors with <code>ruv.norm=TRUE, ruv.k=2</code>.</p>
</li>
</ol>
<blockquote>
<p>RUVse is a modification of RUVs. It estimated the factors of unwanted
variation based on negative control genes from the replicate samples in
each assay group (i.e., enrichment and input), for which the enrichment
effect was assumed to be constant.</p>
</blockquote>
<p>All applied normalization methods can be assessed by
<code>listNormalization</code></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listNormalization.html">listNormalization</a></span><span class="op">(</span><span class="va">Enone</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "TC"                  "UQ"                  "TMM"                </span></span>
<span><span class="co">#&gt;  [4] "DESeq"               "PossionSeq"          "Raw"                </span></span>
<span><span class="co">#&gt;  [7] "Raw_RUVg_k1"         "Raw_RUVg_k2"         "Raw_RUVg_k3"        </span></span>
<span><span class="co">#&gt; [10] "Raw_RUVs_k1"         "Raw_RUVs_k2"         "Raw_RUVs_k3"        </span></span>
<span><span class="co">#&gt; [13] "Raw_RUVse_k1"        "Raw_RUVse_k2"        "Raw_RUVse_k3"       </span></span>
<span><span class="co">#&gt; [16] "TC_RUVg_k1"          "TC_RUVg_k2"          "TC_RUVg_k3"         </span></span>
<span><span class="co">#&gt; [19] "TC_RUVs_k1"          "TC_RUVs_k2"          "TC_RUVs_k3"         </span></span>
<span><span class="co">#&gt; [22] "TC_RUVse_k1"         "TC_RUVse_k2"         "TC_RUVse_k3"        </span></span>
<span><span class="co">#&gt; [25] "UQ_RUVg_k1"          "UQ_RUVg_k2"          "UQ_RUVg_k3"         </span></span>
<span><span class="co">#&gt; [28] "UQ_RUVs_k1"          "UQ_RUVs_k2"          "UQ_RUVs_k3"         </span></span>
<span><span class="co">#&gt; [31] "UQ_RUVse_k1"         "UQ_RUVse_k2"         "UQ_RUVse_k3"        </span></span>
<span><span class="co">#&gt; [34] "TMM_RUVg_k1"         "TMM_RUVg_k2"         "TMM_RUVg_k3"        </span></span>
<span><span class="co">#&gt; [37] "TMM_RUVs_k1"         "TMM_RUVs_k2"         "TMM_RUVs_k3"        </span></span>
<span><span class="co">#&gt; [40] "TMM_RUVse_k1"        "TMM_RUVse_k2"        "TMM_RUVse_k3"       </span></span>
<span><span class="co">#&gt; [43] "DESeq_RUVg_k1"       "DESeq_RUVg_k2"       "DESeq_RUVg_k3"      </span></span>
<span><span class="co">#&gt; [46] "DESeq_RUVs_k1"       "DESeq_RUVs_k2"       "DESeq_RUVs_k3"      </span></span>
<span><span class="co">#&gt; [49] "DESeq_RUVse_k1"      "DESeq_RUVse_k2"      "DESeq_RUVse_k3"     </span></span>
<span><span class="co">#&gt; [52] "PossionSeq_RUVg_k1"  "PossionSeq_RUVg_k2"  "PossionSeq_RUVg_k3" </span></span>
<span><span class="co">#&gt; [55] "PossionSeq_RUVs_k1"  "PossionSeq_RUVs_k2"  "PossionSeq_RUVs_k3" </span></span>
<span><span class="co">#&gt; [58] "PossionSeq_RUVse_k1" "PossionSeq_RUVse_k2" "PossionSeq_RUVse_k3"</span></span></code></pre></div>
<p>Normalized counts can be assessed by <code>Counts</code></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/Counts.html">Counts</a></span><span class="op">(</span><span class="va">Enone</span>, slot<span class="op">=</span><span class="st">"sample"</span>, method<span class="op">=</span><span class="st">"DESeq_RUVg_k2"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                        D1        D2        D3        D4        D5</span></span>
<span><span class="co">#&gt; ENSG00000227232  8.513936 11.645203 12.431441  7.453572 11.104087</span></span>
<span><span class="co">#&gt; ENSG00000268903  3.305409  8.100305  5.277670 26.325845  7.088225</span></span>
<span><span class="co">#&gt; ENSG00000269981  2.270299  4.876642  2.402698 18.118129  6.068101</span></span>
<span><span class="co">#&gt; ENSG00000279457 11.895695  9.998730  8.760526 11.204313 16.375896</span></span>
<span><span class="co">#&gt; ENSG00000225630 13.767575 16.292915 15.365490 10.140265 17.834578</span></span>
<span><span class="co">#&gt; ENSG00000237973 28.042040 37.277287 39.847772 34.078301 33.952026</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="evaluation">Evaluation<a class="anchor" aria-label="anchor" href="#evaluation"></a>
</h4>
<p>To evaluate the performance of normalization, <code>enONE</code>
leverages eight normalization performance metrics that related to
different aspects of the distribution of gene expression measures. The
eight metrics are listed as below:</p>
<ol style="list-style-type: decimal">
<li><p><code>BIO_SIM</code>: Similarity of biological groups. The
average silhouette width of clusters defined by <code>bio.group</code>,
computed with the Euclidean distance metric over the first 3 expression
PCs (default). Large values of <code>BIO_SIM</code> is
desirable.</p></li>
<li><p><code>EN_SIM</code>: Similarity of enrichment groups. The average
silhouette width of clusters defined by <code>en.group</code>, computed
with the Euclidean distance metric over the first 3 expression PCs
(default). Large values of <code>EN_SIM</code> is desirable.</p></li>
<li><p><code>BAT_SIM</code>: Similarity of batch groups. The average
silhouette width of clusters defined by <code>batch.group</code>,
computed with the Euclidean distance metric over the first 3 expression
PCs (default). Low values of <code>BAT_SIM</code> is desirable.</p></li>
<li><p><code>PAM_SIM</code>: Similarity of PAM clustering groups. The
maximum average silhouette width of clusters defined by PAM clustering
(cluster by <code>eval.pam.k</code>), computed with the Euclidean
distance metric over the first 3 expression PCs (default). Large values
of <code>BAT_SIM</code> is desirable.</p></li>
<li><p><code>WV_COR</code>: Preservation of biological variation.
R<sup>2</sup> measure for regression of first 3 expression PCs on first
<code>eval.pc.n</code> PCs of the positive evaluation genes
(<code>PosEvaluation</code>) sub-matrix of the raw count. Large values
of <code>WV_COR</code> is desirable.</p></li>
<li><p><code>UV_COR</code>: Removal of unwanted variation. R<sup>2</sup>
measure for regression of first 3 expression PCs on first
<code>eval.pc.n</code> PCs of the negative evaluation genes
(<code>NegEvaluation</code>) sub-matrix of the raw count. Low values of
<code>UV_COR</code> is desirable.</p></li>
<li><p><code>RLE_MED</code>: The mean squared-median Relative Log
Expression (RLE). Low values of <code>RLE_MED</code> is
desirable.</p></li>
<li><p><code>RLE_IQR</code>: The variance of inter-quartile range (IQR)
of RLE. Low values of <code>RLE_IQR</code> is desirable.</p></li>
</ol>
<p>Evaluation metrics can be assessed by <code>getMetrics</code></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getMetrics.html">getMetrics</a></span><span class="op">(</span><span class="va">Enone</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span></span>
<span><span class="co">#&gt;                        BIO_SIM    EN_SIM   BATCH_SIM   PAM_SIM      RLE_MED</span></span>
<span><span class="co">#&gt; DESeq_RUVg_k2      -0.02467593 0.3576650 -0.03483061 0.5713398 9.319387e-06</span></span>
<span><span class="co">#&gt; PossionSeq_RUVs_k3  0.10240479 0.3374029 -0.03965069 0.5556398 9.971803e-05</span></span>
<span><span class="co">#&gt; TMM_RUVg_k2        -0.02664678 0.3518415 -0.03524885 0.5595637 8.500478e-05</span></span>
<span><span class="co">#&gt; DESeq_RUVg_k3      -0.04651463 0.3485155 -0.04427468 0.5708689 1.370353e-05</span></span>
<span><span class="co">#&gt; DESeq_RUVs_k1       0.03352455 0.3391415 -0.04444996 0.5224180 7.861089e-06</span></span>
<span><span class="co">#&gt;                        RLE_IQR    WV_COR    UV_COR</span></span>
<span><span class="co">#&gt; DESeq_RUVg_k2      0.011996148 0.7401496 0.3258051</span></span>
<span><span class="co">#&gt; PossionSeq_RUVs_k3 0.012314397 0.7709943 0.3154878</span></span>
<span><span class="co">#&gt; TMM_RUVg_k2        0.012089861 0.7404610 0.3288269</span></span>
<span><span class="co">#&gt; DESeq_RUVg_k3      0.011133971 0.6688187 0.4166522</span></span>
<span><span class="co">#&gt; DESeq_RUVs_k1      0.009145155 0.6495402 0.3680677</span></span></code></pre></div>
<p>Evaluation score is the average rank of each performance metrics,
which can be assessed by <code>getScore</code></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getScore.html">getScore</a></span><span class="op">(</span><span class="va">Enone</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span></span>
<span><span class="co">#&gt;                    BIO_SIM EN_SIM BATCH_SIM PAM_SIM RLE_MED RLE_IQR WV_COR</span></span>
<span><span class="co">#&gt; DESeq_RUVg_k2           32     59        40      59      56      25     20</span></span>
<span><span class="co">#&gt; PossionSeq_RUVs_k3      60     23        47      50      33      23     23</span></span>
<span><span class="co">#&gt; TMM_RUVg_k2             30     50        42      54      39      24     21</span></span>
<span><span class="co">#&gt; DESeq_RUVg_k3           20     48        53      58      54      29     12</span></span>
<span><span class="co">#&gt; DESeq_RUVs_k1           55     32        54      25      57      32      5</span></span>
<span><span class="co">#&gt;                    UV_COR  SCORE</span></span>
<span><span class="co">#&gt; DESeq_RUVg_k2          59 43.750</span></span>
<span><span class="co">#&gt; PossionSeq_RUVs_k3     60 39.875</span></span>
<span><span class="co">#&gt; TMM_RUVg_k2            58 39.750</span></span>
<span><span class="co">#&gt; DESeq_RUVg_k3          39 39.125</span></span>
<span><span class="co">#&gt; DESeq_RUVs_k1          52 39.000</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="select-the-suitable-normalization-for-subsequent-analysis">Select the suitable normalization for subsequent analysis<a class="anchor" aria-label="anchor" href="#select-the-suitable-normalization-for-subsequent-analysis"></a>
</h2>
<p><code>enONE</code> provides biplot to exploit the full space of
normalization methods, you can turn on the interactive mode with
<code>interactive=TRUE</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get performance score</span></span>
<span><span class="va">enScore</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getScore.html">getScore</a></span><span class="op">(</span><span class="va">Enone</span><span class="op">)</span></span>
<span><span class="co"># perform PCA based on evaluation score, excluding BAT_SIM column (3) if no batch information provided, and SCORE column (9).</span></span>
<span><span class="co"># pca.eval &lt;- prcomp(enScore[,-c(3, 9)], scale = TRUE)</span></span>
<span><span class="va">pca.eval</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="va">enScore</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">9</span><span class="op">)</span><span class="op">]</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># pca biplot</span></span>
<span><span class="fu"><a href="../reference/PCA_Biplot.html">PCA_Biplot</a></span><span class="op">(</span><span class="va">pca.eval</span>, score <span class="op">=</span> <span class="va">enScore</span><span class="op">$</span><span class="va">SCORE</span>, interactive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="enONE_tutorial_files/figure-html/unnamed-chunk-12-1.png" width="960"></p>
<p>In this plot, each point corresponds to a normalization procedure and
is colored by the performance score (mean of eight scone performance
metric ranks). The blue arrows correspond to the PCA loadings for the
performance metrics. The direction and length of a blue arrow can be
interpreted as a measure of how much each metric contributed to the
first two PCs</p>
<p>Here, we use the top-ranked procedure DESeq_RUVg_k2 for downstream
analysis.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select normalization</span></span>
<span><span class="va">norm.method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">enScore</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get normalized counts</span></span>
<span><span class="va">norm.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Counts.html">Counts</a></span><span class="op">(</span><span class="va">Enone</span>, slot <span class="op">=</span> <span class="st">"sample"</span>, method <span class="op">=</span> <span class="va">norm.method</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get normalization factors</span></span>
<span><span class="va">norm.factors</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getFactor.html">getFactor</a></span><span class="op">(</span><span class="va">Enone</span>, slot <span class="op">=</span> <span class="st">"sample"</span>, method <span class="op">=</span> <span class="va">norm.method</span><span class="op">)</span></span>
<span></span>
<span><span class="va">norm.method</span></span>
<span><span class="co">#&gt; [1] "DESeq_RUVg_k2"</span></span></code></pre></div>
<blockquote>
<p>To be noted, if normalized counts are not returned in
<code>enONE</code> run step (i.e., <code>return.norm=FALSE</code>), you
can manually perform the normalization method, e.g.,</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># perform normalization</span></span>
<span><span class="va">Enone</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/UseNormalization.html">UseNormalization</a></span><span class="op">(</span><span class="va">Enone</span>, slot <span class="op">=</span> <span class="st">"sample"</span>, method <span class="op">=</span> <span class="va">norm.method</span><span class="op">)</span></span>
<span><span class="co"># get normalized counts</span></span>
<span><span class="va">norm.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Counts.html">Counts</a></span><span class="op">(</span><span class="va">Enone</span>, slot <span class="op">=</span> <span class="st">"sample"</span>, method <span class="op">=</span> <span class="va">norm.method</span><span class="op">)</span></span></code></pre></div>
</blockquote>
</div>
<div class="section level2">
<h2 id="effect-of-normalization">Effect of normalization<a class="anchor" aria-label="anchor" href="#effect-of-normalization"></a>
</h2>
<p>We perform PCA based on the count matrix from sample of interest
before and after the normalization, for demonstrating the effect of
normalization.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create sample name, e.g., Y3.Input</span></span>
<span><span class="va">samples_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">Enone</span><span class="op">$</span><span class="va">condition</span>, <span class="va">Enone</span><span class="op">$</span><span class="va">replicate</span>, sep<span class="op">=</span><span class="st">"."</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># PCA for raw count</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PCAplot.html">PCAplot</a></span><span class="op">(</span><span class="fu"><a href="../reference/Counts.html">Counts</a></span><span class="op">(</span><span class="va">Enone</span>, slot<span class="op">=</span><span class="st">"sample"</span>, <span class="st">"Raw"</span><span class="op">)</span>, </span>
<span>            color <span class="op">=</span> <span class="va">Enone</span><span class="op">$</span><span class="va">batch</span>,</span>
<span>            shape <span class="op">=</span> <span class="va">Enone</span><span class="op">$</span><span class="va">enrich</span>,</span>
<span>            label <span class="op">=</span> <span class="va">samples_name</span>, </span>
<span>            vst.norm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Before normalization"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># PCA for normalized count</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PCAplot.html">PCAplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">norm.data</span><span class="op">)</span>, </span>
<span>            color <span class="op">=</span> <span class="va">Enone</span><span class="op">$</span><span class="va">batch</span>,</span>
<span>            shape <span class="op">=</span> <span class="va">Enone</span><span class="op">$</span><span class="va">enrich</span>,</span>
<span>            label <span class="op">=</span> <span class="va">samples_name</span>, </span>
<span>            vst.norm <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"After normalization"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># combine two plots</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span>
<span><span class="co">#&gt; Warning: ggrepel: 1 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">#&gt; increasing max.overlaps</span></span></code></pre></div>
<p><img src="enONE_tutorial_files/figure-html/unnamed-chunk-15-1.png" width="1152"></p>
</div>
<div class="section level2">
<h2 id="find-enrichment">Find enrichment<a class="anchor" aria-label="anchor" href="#find-enrichment"></a>
</h2>
<p><code>enONE</code> package can help you find enrichment genes from
each biological groups via differential expression. It can identify
genes that significantly increased in enrichment samples compared to
input samples.</p>
<p><code>FindEnrichment</code> automates this processes for all groups
provided in <code>bio.group</code>.</p>
<p>By default, enriched genes (NAD-RNA in here) are defined as fold
change of normalized transcript counts ≥ 2
(<code>logfc.cutoff = 1</code>), FDR &lt; 0.05
(<code>p.cutoff = 0.05</code>) in enrichment samples compared to those
in input samples.</p>
<p>Use <code>getEnrichment</code> to retrieve a list of enrichment
result tables.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># find all enriched genes</span></span>
<span><span class="va">Enone</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindEnrichment.html">FindEnrichment</a></span><span class="op">(</span><span class="va">Enone</span>, slot<span class="op">=</span><span class="st">"sample"</span>, norm.method <span class="op">=</span> <span class="va">norm.method</span>, </span>
<span>                        logfc.cutoff <span class="op">=</span> <span class="fl">1</span>, p.cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">#&gt; - Estimate dispersion &amp; Fit GLM... </span></span>
<span><span class="co">#&gt; - Testing differential genes...</span></span>
<span></span>
<span><span class="co"># get filtered enrichment results</span></span>
<span><span class="va">res.sig.ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getEnrichment.html">getEnrichment</a></span><span class="op">(</span><span class="va">Enone</span>, slot<span class="op">=</span><span class="st">"sample"</span>, filter<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># count number of enrichment in each group</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">res.sig.ls</span>, <span class="va">nrow</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Y.Enrich_Y.Input M.Enrich_M.Input O.Enrich_O.Input </span></span>
<span><span class="co">#&gt;              579              672              683</span></span></code></pre></div>
<p>Each enrichment table is a <code>data.frame</code> with a list of
genes as rows, and associated information as columns (GeneID, logFC,
p-values, etc.). The following columns are present in the table:</p>
<ul>
<li>
<code>GeneID</code>: ID of genes.</li>
<li>
<code>logFC</code>: log2 fold-change between enrichment and input
samples. Positive values indicate that the gene is more highly enriched
in the enrichment group.</li>
<li>
<code>logCPM</code>: log2 CPM (counts per million) of the average
expression of all samples.</li>
<li>
<code>LR</code>: Likelihood ratio of the likelihood ratio test.</li>
<li>
<code>PValue</code>: p-value from the likelihood ratio test.</li>
<li>
<code>FDR</code>: False discovery rate of the p-value, default “BH”
method is applied.</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res.sig.ls</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;            GeneID    logFC   logCPM       LR        PValue           FDR</span></span>
<span><span class="co">#&gt; 1 ENSG00000013275 2.271643 7.518198 731.6422 3.936608e-161 5.444329e-157</span></span>
<span><span class="co">#&gt; 2 ENSG00000113387 2.643018 9.732157 631.6712 2.164465e-139 1.496728e-135</span></span>
<span><span class="co">#&gt; 3 ENSG00000104853 2.244442 7.839256 587.7484 7.738626e-130 3.567507e-126</span></span>
<span><span class="co">#&gt; 4 ENSG00000105185 2.094086 6.147852 558.6392 1.662061e-123 5.746574e-120</span></span>
<span><span class="co">#&gt; 5 ENSG00000143110 3.084209 9.115976 550.0416 1.232914e-121 3.410239e-118</span></span>
<span><span class="co">#&gt; 6 ENSG00000173915 2.324782 7.005478 483.7653 3.239173e-107 7.466295e-104</span></span></code></pre></div>
<p>To convert list of enrichment result tables into data frame in long
format, <code>enONE</code> package provide <code>reduceRes</code>
function for this task. Finally, you can use
<code>BetweenStatPlot</code> to compare the global extent of NAD-RNA
modification level between groups.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># simplify group id</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res.sig.ls</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Young"</span>, <span class="st">"Mid"</span>, <span class="st">"Old"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># logfc.col specify the name of logFC column</span></span>
<span><span class="va">nad_df1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/reduceRes.html">reduceRes</a></span><span class="op">(</span><span class="va">res.sig.ls</span>, logfc.col <span class="op">=</span> <span class="st">"logFC"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># convert the Group column as factor</span></span>
<span><span class="va">nad_df1</span><span class="op">$</span><span class="va">Group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">nad_df1</span><span class="op">$</span><span class="va">Group</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">nad_df1</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># draw plot</span></span>
<span><span class="va">bxp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/BetweenStatPlot.html">BetweenStatPlot</a></span><span class="op">(</span><span class="va">nad_df1</span>, x<span class="op">=</span><span class="st">"Group"</span>, y<span class="op">=</span><span class="st">"logFC"</span>, color<span class="op">=</span><span class="st">"Group"</span>, </span>
<span>                        step.increase <span class="op">=</span> <span class="fl">0.6</span>, add.p <span class="op">=</span> <span class="st">"p"</span>, </span>
<span>                        comparisons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Young"</span>, <span class="st">"Mid"</span><span class="op">)</span>, </span>
<span>                                           <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Young"</span>, <span class="st">"Old"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bxp1</span></span></code></pre></div>
<p><img src="enONE_tutorial_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="handling-synthetic-spike-in-rna">Handling synthetic spike-in RNA<a class="anchor" aria-label="anchor" href="#handling-synthetic-spike-in-rna"></a>
</h2>
<p>Since synthetic RNAs, of which one with 5% NAD-caps and another with
100% m7G-caps, are included, we can use these spike-ins to determine the
capture sensitivity and specificity.</p>
<p><code>synEnrichment</code> calculate the enrichment levels of
synthetic spike-ins with given normalization method.
<code>DotPlot</code> can be used to visualize the enrichment levels of
synthetic spike-in.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compute synthetic spike-in enrichment</span></span>
<span><span class="va">syn_level</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/synEnrichment.html">synEnrichment</a></span><span class="op">(</span><span class="va">Enone</span>, method<span class="op">=</span><span class="va">norm.method</span>, log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># transform to long format</span></span>
<span><span class="va">syn_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">syn_level</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="st">"syn_id"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">syn_id</span>,</span>
<span>               names_to <span class="op">=</span> <span class="st">"id"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"logFC"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">meta</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"id"</span>,<span class="st">"condition"</span><span class="op">)</span><span class="op">]</span>, by<span class="op">=</span><span class="st">"id"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove suffix of condition for simplification</span></span>
<span><span class="va">syn_df</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\..*"</span>, <span class="st">""</span>, <span class="va">syn_df</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># rename facet label</span></span>
<span><span class="va">samples_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"5% NAD-RNA"</span>, <span class="st">"100% m7G-RNA"</span><span class="op">)</span>, </span>
<span>                          nm<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Syn1"</span>, <span class="st">"Syn2"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># draw dotplot</span></span>
<span><span class="fu"><a href="../reference/DotPlot.html">DotPlot</a></span><span class="op">(</span><span class="va">syn_df</span>, x<span class="op">=</span><span class="st">"syn_id"</span>, y<span class="op">=</span><span class="st">"logFC"</span>, fill<span class="op">=</span><span class="st">"syn_id"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_x_discrete</a></span><span class="op">(</span>labels<span class="op">=</span><span class="va">samples_label</span><span class="op">)</span></span>
<span><span class="co">#&gt; Bin width defaults to 1/30 of the range of the data. Pick better value with</span></span>
<span><span class="co">#&gt; `binwidth`.</span></span></code></pre></div>
<p><img src="enONE_tutorial_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>Syn1, which contained 5% NAD-RNA, are significantly enriched, whereas
no enrichment is found for Syn2 made up with 100% m7G-RNA.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># save Enone data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">Enone</span>, file<span class="op">=</span><span class="st">"data/Enone.RData"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] patchwork_1.1.2 lubridate_1.9.2 forcats_1.0.0   stringr_1.5.0  </span></span>
<span><span class="co">#&gt;  [5] dplyr_1.1.2     purrr_1.0.1     readr_2.1.4     tidyr_1.3.0    </span></span>
<span><span class="co">#&gt;  [9] tibble_3.2.1    ggplot2_3.4.2   tidyverse_2.0.0 enONE_0.1.0    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] bitops_1.0-7                pbapply_1.7-0              </span></span>
<span><span class="co">#&gt;   [3] rlang_1.1.1                 magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;   [5] matrixStats_0.63.0          compiler_4.3.0             </span></span>
<span><span class="co">#&gt;   [7] flexmix_2.3-19              systemfonts_1.0.4          </span></span>
<span><span class="co">#&gt;   [9] vctrs_0.6.2                 pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [11] crayon_1.5.2                fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [13] backports_1.4.1             XVector_0.40.0             </span></span>
<span><span class="co">#&gt;  [15] labeling_0.4.2              utf8_1.2.3                 </span></span>
<span><span class="co">#&gt;  [17] rmarkdown_2.21              tzdb_0.4.0                 </span></span>
<span><span class="co">#&gt;  [19] ragg_1.2.5                  xfun_0.39                  </span></span>
<span><span class="co">#&gt;  [21] modeltools_0.2-23           zlibbioc_1.46.0            </span></span>
<span><span class="co">#&gt;  [23] cachem_1.0.8                GenomeInfoDb_1.36.0        </span></span>
<span><span class="co">#&gt;  [25] jsonlite_1.8.4              EnvStats_2.7.0             </span></span>
<span><span class="co">#&gt;  [27] highr_0.10                  DelayedArray_0.26.2        </span></span>
<span><span class="co">#&gt;  [29] fpc_2.2-10                  BiocParallel_1.34.1        </span></span>
<span><span class="co">#&gt;  [31] broom_1.0.4                 parallel_4.3.0             </span></span>
<span><span class="co">#&gt;  [33] prabclus_2.3-2              cluster_2.1.4              </span></span>
<span><span class="co">#&gt;  [35] R6_2.5.1                    bslib_0.4.2                </span></span>
<span><span class="co">#&gt;  [37] stringi_1.7.12              limma_3.56.1               </span></span>
<span><span class="co">#&gt;  [39] car_3.1-2                   GenomicRanges_1.52.0       </span></span>
<span><span class="co">#&gt;  [41] jquerylib_0.1.4             diptest_0.76-0             </span></span>
<span><span class="co">#&gt;  [43] Rcpp_1.0.10                 SummarizedExperiment_1.30.1</span></span>
<span><span class="co">#&gt;  [45] knitr_1.42                  IRanges_2.34.0             </span></span>
<span><span class="co">#&gt;  [47] splines_4.3.0               timechange_0.2.0           </span></span>
<span><span class="co">#&gt;  [49] Matrix_1.5-4                nnet_7.3-18                </span></span>
<span><span class="co">#&gt;  [51] tidyselect_1.2.0            abind_1.4-5                </span></span>
<span><span class="co">#&gt;  [53] yaml_2.3.7                  codetools_0.2-19           </span></span>
<span><span class="co">#&gt;  [55] lattice_0.21-8              withr_2.5.0                </span></span>
<span><span class="co">#&gt;  [57] Biobase_2.60.0              evaluate_0.21              </span></span>
<span><span class="co">#&gt;  [59] desc_1.4.2                  mclust_6.0.0               </span></span>
<span><span class="co">#&gt;  [61] kernlab_0.9-32              pillar_1.9.0               </span></span>
<span><span class="co">#&gt;  [63] ggpubr_0.6.0                MatrixGenerics_1.12.0      </span></span>
<span><span class="co">#&gt;  [65] carData_3.0-5               stats4_4.3.0               </span></span>
<span><span class="co">#&gt;  [67] plotly_4.10.1               generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [69] rprojroot_2.0.3             RCurl_1.98-1.12            </span></span>
<span><span class="co">#&gt;  [71] hms_1.1.3                   S4Vectors_0.38.1           </span></span>
<span><span class="co">#&gt;  [73] munsell_0.5.0               scales_1.2.1               </span></span>
<span><span class="co">#&gt;  [75] class_7.3-21                glue_1.6.2                 </span></span>
<span><span class="co">#&gt;  [77] lazyeval_0.2.2              tools_4.3.0                </span></span>
<span><span class="co">#&gt;  [79] robustbase_0.95-1           data.table_1.14.8          </span></span>
<span><span class="co">#&gt;  [81] locfit_1.5-9.7              ggsignif_0.6.4             </span></span>
<span><span class="co">#&gt;  [83] fs_1.6.2                    grid_4.3.0                 </span></span>
<span><span class="co">#&gt;  [85] edgeR_3.42.2                colorspace_2.1-0           </span></span>
<span><span class="co">#&gt;  [87] GenomeInfoDbData_1.2.10     cli_3.6.1                  </span></span>
<span><span class="co">#&gt;  [89] textshaping_0.3.6           fansi_1.0.4                </span></span>
<span><span class="co">#&gt;  [91] viridisLite_0.4.2           S4Arrays_1.0.4             </span></span>
<span><span class="co">#&gt;  [93] paintingr_0.1.0             gtable_0.3.3               </span></span>
<span><span class="co">#&gt;  [95] DEoptimR_1.0-13             rstatix_0.7.2              </span></span>
<span><span class="co">#&gt;  [97] DESeq2_1.40.1               sass_0.4.6                 </span></span>
<span><span class="co">#&gt;  [99] digest_0.6.31               BiocGenerics_0.46.0        </span></span>
<span><span class="co">#&gt; [101] ggrepel_0.9.3               farver_2.1.1               </span></span>
<span><span class="co">#&gt; [103] htmlwidgets_1.6.2           memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [105] htmltools_0.5.5             pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt; [107] lifecycle_1.0.3             httr_1.4.6                 </span></span>
<span><span class="co">#&gt; [109] MASS_7.3-58.4</span></span></code></pre></div>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Dean Li.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
